### PR TITLE
PUBDEV-8936 Shapley bugfix for DRF

### PR DESCRIPTION
**Pull request by valenad1**
https://h2oai.atlassian.net/browse/PUBDEV-8936

### CODE REVIEW

**Comment by valenad1** [comment]
Hi @navdeep-G,

Can you help me to validate this bugfix please, if you still remember anything from Shapley implementations?

User found a bug in the shapley implementation for DRF binary classification.

Problem: When the feature is not used at all then the contribution is not zero:


```
import h2o
from h2o.estimators import H2ORandomForestEstimator

prostate = h2o.import_file("http://s3.amazonaws.com/h2o-public-test-data/smalldata/prostate/prostate.csv")
prostate["CAPSULE"] = prostate["CAPSULE"].asfactor()
predictors = ["AGE","RACE","DPROS","DCAPS","PSA","VOL","GLEASON"]
response = "CAPSULE"

pros_rf = H2ORandomForestEstimator(ntrees=1,
                                   max_depth= 1,
                                   seed=1111)

pros_rf.train(y = response, training_frame = prostate)
display(pros_rf.predict_contributions(prostate))
pros_rf.varimp(use_pandas=True)
```

gives

<div class="output_area"><div class="output_subarea output_html rendered_html" dir="auto">

ID | AGE | RACE | DPROS | DCAPS | PSA | VOL | GLEASON | BiasTerm
-- | -- | -- | -- | -- | -- | -- | -- | --
0.111111 | 0.111111 | 0.111111 | 0.111111 | 0.0683616 | 0.111111 | 0.111111 | 0.111111 | -0.486257
0.111111 | 0.111111 | 0.111111 | 0.111111 | 0.464577 | 0.111111 | 0.111111 | 0.111111 | -0.486257
0.111111 | 0.111111 | 0.111111 | 0.111111 | 0.464577 | 0.111111 | 0.111111 | 0.111111 | -0.486257
0.111111 | 0.111111 | 0.111111 | 0.111111 | 0.0683616 | 0.111111 | 0.111111 | 0.111111 | -0.486257
0.111111 | 0.111111 | 0.111111 | 0.111111 | 0.0683616 | 0.111111 | 0.111111 | 0.111111 | -0.486257
0.111111 | 0.111111 | 0.111111 | 0.111111 | 0.464577 | 0.111111 | 0.111111 | 0.111111 | -0.486257
0.111111 | 0.111111 | 0.111111 | 0.111111 | 0.464577 | 0.111111 | 0.111111 | 0.111111 | -0.486257
0.111111 | 0.111111 | 0.111111 | 0.111111 | 0.464577 | 0.111111 | 0.111111 | 0.111111 | -0.486257
0.111111 | 0.111111 | 0.111111 | 0.111111 | 0.0683616 | 0.111111 | 0.111111 | 0.111111 | -0.486257
0.111111 | 0.111111 | 0.111111 | 0.111111 | 0.464577 | 0.111111 | 0.111111 | 0.111111 | -0.486257

</div>



  | variable | relative_importance | scaled_importance | percentage
-- | -- | -- | -- | --
DCAPS | 5.741998 | 1.0 | 1.0
ID | 0.000000 | 0.0 | 0.0
AGE | 0.000000 | 0.0 | 0.0
RACE | 0.000000 | 0.0 | 0.0
DPROS | 0.000000 | 0.0 | 0.0
PSA | 0.000000 | 0.0 | 0.0
VOL | 0.000000 | 0.0 | 0.0
GLEASON | 0.000000 | 0.0 | 0.0

I think the problem is [here](https://github.com/h2oai/h2o-3/pull/6452/files#diff-8c3135e3cd852b8e09bd317098e0e4e85314303152740f67c18411a8fdd6d7f9R123). Can you please confirm that my assumption is correct? We should "normalized" only when the feature is not zero.

Thank you!

**Review by navdeep-G — APPROVED** [review]
This LGTM.

**Comment by navdeep-G** [comment]
@valenad1 This looks like the right fix to me. I just approved.

**Review by wendycwong — APPROVED** [review]

File: h2o-algos/src/main/java/hex/tree/drf/DRFModel.java [hunk 1]
```diff
@@ -120,8 +120,8 @@ public void addContribToNewChunk(float[] contribs, NewChunk[] nc) {
             if (_output.nclasses() == 1) { //Regression
                 nc[i].addNum(contribs[i] /_output._ntrees);
             } else { //Binomial
-                float featurePlusBiasRatio = (float)1 / (_output.nfeatures() + 1); // + 1 for bias term
-                nc[i].addNum(featurePlusBiasRatio - (contribs[i] / _output._ntrees));
+              float featurePlusBiasRatio = (float)1 / (_output._varimp.numberOfUsedVariables() + 1); // + 1 for bias term
+              nc[i].addNum(contribs[i] != 0 ? (featurePlusBiasRatio - (contribs[i] / _output._ntrees)) : (contribs[i] / _output._ntrees));
```
[the remark this answers is not in the document]
Comment (navdeep-G) [reply h2o-algos/src/main/java/hex/tree/drf/DRFModel.java:124]: This should evaluate to `0` if the contrib is not `0`, i.e., `contrib[i] == 0`. The edge case, which is unlikely, is the contrib being negative.

[the remark this answers is not in the document]
Comment (valenad1) [reply h2o-algos/src/main/java/hex/tree/drf/DRFModel.java]: Fixed, thank you!

File: h2o-algos/src/test/java/hex/tree/drf/DRFPredictContribsTest.java [hunk 2]
```diff
@@ -21,243 +24,254 @@
...
-            Frame adapted = new Frame(fr);
-            drf.adaptTestForTrain(adapted, true, false);
-
-            for (int i = 0; i < parms._ntrees; i++) {
-                new hex.tree.drf.DRFPredictContribsTest.CheckTreeSHAPTask(drf, i).doAll(adapted);
-            }
-        } finally {
-            Scope.exit();
+        @BeforeClass
+        public static void stall() {
```
[the remark this answers is not in the document]
Comment (valenad1) [reply h2o-algos/src/test/java/hex/tree/drf/DRFPredictContribsTest.java]: I am trying to make commits atomic as much as possible to help guide reviewer :)

**Review by mn-mikke — APPROVED** [review]
LGTM, thank you!

**Pull request merged**